### PR TITLE
Add resticw (restic wrapper) utility

### PR DESCRIPTION
**Pull request by gerardbosch**
The script provides a convenient way to load environment config, deal
with profiles and act as a pass-through to restic. The overall thing is
to improve the UX when running restic, integrating the features this
project provides.

## Note

The script itself is a very simple thing.
The command line parser is auto-generated using [docopt.sh](https://github.com/andsens/docopt.sh) driven from
the script's DOC. It can be refreshed upon DOC changes with:
`docopt.sh path/to/resticw`.

### Examples

```console
resticw stats latest

resticw -p profileA snapshots
```

(No need of sourcing any file, works on any spawned terminal)

### Help

```console
❯ resticw --help
A little wrapper over restic just to handle profiles and environment loading.

  It loads the backup profile/environment in a subshell to avoid any credential leak (Note: Run it with sudo so it can load the environment).

  Usage:
    resticw [options] <restic_arguments>

    The restic_arguments is just the regular unwrapped restic arguments, e.g. stats latest

  Options:
    -p --profile=<name>        Specify the profile to load or use default [default: default].

  Examples:
    resticw --profile profileA snapshots
    resticw stats latest  # this will use the profile: default
```

### CODE REVIEW

File: Makefile
Comment (gerardbosch) [inline Makefile:35]: Using 755 here as, the user should be able to find `resticw` discoverable, but by using 744, `which resticw` shows nothing, as well as tab  auto-completion does not help when typing the command.

**Comment by gerardbosch** [comment]
The parser could scare, but as it is auto-generated it does not require the burden of manual manteinance, just issue a command. I always found CLI parsing tough, hehe, and found docopt.sh very nice.

**Comment by erikw** [comment]
I just sneak-peaked a bit and it looks interesting!

I make a proper review and try-out later in the week :)

File: README.md
Comment (erikw) [inline README.md:78]: Thanks for writing REAME docs ⭐ 

To keep the base install minimal,  simple, easy to follow, let's move this to after "9. Optional: automated backup checks" as "10. Optional: restic profile wrapper" or something like this.

Later in  #58 I will restructure the optionals better in the README

**Comment by gerardbosch** [comment]
Sure! There is one part of the update mechanism implemented as a separate bash script, as Makefile started to get convoluted and limitating.

These changes will require adjustments as per the recently added linter. But I keep it on hold for a while :)

File: README.md
Comment (erikw) [inline README.md:95]: To make this easy to maintain, let's add a bullet point in the section `# Development`, that lists the minimal commands to update the parser

```console
$ pip install doctopt.sh
$ doctopt.sh usr/local/sbin/resticw
```

or something. It's very useful to have these short command notes down the line when adding a new paramter in a few years.. :)

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:28]: I think docopt looks cook, but it's sure overkill for this use case of having just one option. Just checking manually `$1`, or using the standard built-in `getopts` (see my template here https://github.com/erikw/templates/blob/master/bash/getopts.sh#L30-L38) would suffice.

But programming should also be fun, and trying out new techs is fun,... so let's give it a try :)

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:131]: 👍 for unique exit codes

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:133]: Hmm I would disagree actually, this project need actually not be run by root. If one installs it in a directory hierarchy owned by root, then yes.

For the macOS setup that will be added, the project will typically be installed in `/usr/local/` where normal users have write access (this is where the Homebrew package manager installed applications by default). This is why #49 is important to fix.

So I'd say let's remove this actually. It should be obvious if one is not running as the right user

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:138]: Tip: use `printf(1)` for more advanced formatting.

```bash
printf "‣ Using profile: %s -- (%s)\n" "${__profile}" "${profile_file}"
```

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:140]: How would credentials be leaked if we don't spawn a subshell? The script itself already runs in a sub-shell from where it was started.

**Review by erikw — CHANGES_REQUESTED** [review]
Cool, I think this looks useful!

I've not actually tried it yet, to be done...

File: usr/local/sbin/resticw
Comment (erikw) [reply usr/local/sbin/resticw:121]: When I run `$ docopt.sh resticw` I get a warning:

```
resticw No invocations of docopt found, check your script to make sure this is correct.
docopt.sh is invoked with `eval "$(docopt "$@")"`.

```

File: README.md
Comment (erikw) [inline README.md:83]: sudo not required as per comment below

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:11]: When just running like `$ resticw` without arguments, the helptext is missing the 2 last characters from this line:

```
   Usage:
      resticw [options] <restic_arguemnt
                                        ^
```

Comment (erikw) [reply usr/local/sbin/resticw:11]: I think restic calls the first level of argument commands, which can have arguments. In `$restic backup /` `backup` would be the command and `/` the argument to that command =>

```
 Usage:
    resticw [restic_options] <restic_command> [restic_command_options]
```

?

**Comment by erikw** [comment]
I've tested the script now. All in all it's great - I will use it myself on my systems :)

Just a few comments above to look at

**Comment by gerardbosch** [comment]
Cool! I try to check the comments, answer the ones I can or address it in a while :D

File: Makefile
Comment (gerardbosch) [reply Makefile:35]: :+1:

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw:123]: I opened few days ago this https://github.com/andsens/docopt.sh/issues/19 regarding a problem with `set -u`. The author answered on that with the `set +u` trick and it's view on that :)

But as you've just faced, there is a weird warning now. I've just written again to the issue to notice that false warning/error.

Comment (gerardbosch) [reply usr/local/sbin/resticw:123]: It looks like a false alarm, as it works fine. I see 3 options:
- Remove the `set -u` from the top of the script and get rid of the `set +u` workaround then.
- Accept and live with that false error when refreshing the parser 😆
- Wait to see if the message gets fixed in origin, maybe less probable

File: README.md
Comment (gerardbosch) [reply README.md:78]: Perfect! No worries :)

File: README.md
Comment (gerardbosch) [reply README.md:95]: Sure, done!

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw:26]: Yeah, hehe. I liked to save myself the burden of writing that tough stuff. And it comes with nice benefits, like short and long options, `--help` and usage out-of-the-box :D  Also good way to try stuff.

Comment (gerardbosch) [reply usr/local/sbin/resticw:26]: Also forces you to write good usage DOC messages, hehe

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw]: Great! Thanks to point that.
I've change it to only issue that message (and error) in case the profile is not readable by the user. Check last version and let me know :)

In any case, of failure, maybe is better to echo a warning and exit, than assume the user remembers about internals and file permissions. Not showing a message would make it fail anyway but maybe less obvious what's the problem.

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw:143]: Hey, thanks! I can't spot much of a difference, the `printf` version being even longer. Do you want to change this?

I find the string interpolation in echo much easier to read than using `%format` and passing actual arguments at the end. It is not using any advanced formating actually, hehe.

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw]: Oh! totally true! I removed the note, and the sub-shell. I think I got confused the first time I wrote this script time ago :) Credentials are confined to the shell spawned by running `resticw` and the environment dies after command finishes.

File: README.md
Comment (gerardbosch) [reply README.md:83]: Updated

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw]: Yep! I realised the missing text on `$ resticw` before, and after comparing with the docopt.sh example, now, I managed to fix it. It should work if you update the branch :)

---
For the other thing of `<restic_command> [restic_command_options]` I tried to change it and just finally reworded as `<restic_arguments_line>...`. I think splitting in 2 (command, options) does not work well with the parser. But it help me to fix something, now there is `...` that was missing to allow multiple tokens. Also, from the shell standpoint, they are the arguments provided to the restic command (it depends on which perspective you observe, hehe).

But the ultimate reason is that it didn't work well.

File: usr/local/sbin/resticw
Comment (erikw) [reply usr/local/sbin/resticw:26]: I don't say no to good documentation!

File: usr/local/sbin/resticw
Comment (erikw) [reply usr/local/sbin/resticw:123]: With the motivation that the acutal scrip starts at line 128, I think it makes sense to move `set -euo pipefail` there actually, and let docopt do its thing before that

```
# --vvv--             ACTUAL SCRIPT BELOW             --vvv--

set -euo pipefail
ENV_DIR=/etc/restic
[...]
```

File: usr/local/sbin/resticw
Comment (erikw) [reply usr/local/sbin/resticw:143]: Preferences of style :). I see benefits in both. No need to fix, just a tip

**Comment by erikw** [comment]
> One thing to ask, I've seen most of the stuff in the project is using tabs instead of spaces. I wrote that one with space indentation, but I can commit to replace it by tabs if you prefer, to be consistent with the rest of the project. I'm not tab fan, but no problem to change it, hehe.

Good point. I was raised with hard tabs, coming from a background C with Linux kernel coding style convention... and I still prefer hard tabs to soft everyday ;P

In the end it really does not matter, as long as the code base is consistent. We should add the `.editorconfig` I saw in your other PR.

For consistency, let's use hard tabs. With `.editorconfig` in place this will be transparent anyways :)

File: usr/local/sbin/resticw
Comment (erikw) [reply usr/local/sbin/resticw]: Now it looks better :)

File: usr/local/sbin/resticw
Comment (erikw) [reply usr/local/sbin/resticw]: That's a great approach, I like it!

Nitpick; technically the error message shoud lnot say  `Error: Run it as root or use sudo` because maybe someone has set up a different use account `restic-account` that is used for backup. The message should be more generic like `Error: can't ready the profile. Are you running this script as the right user? Try su or sudo`

or something like this :)

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:134]: Linter says

```
In /github/workspace/usr/local/sbin/resticw line 134:
profile_file="${ENV_DIR}/${__profile}.env"
                         ^----------^ SC2154 (warning): __profile is referenced but not assigned.
```

I think it's not able to parse the docopt function above. We might need to [ignore](https://github.com/koalaman/shellcheck/wiki/Ignore) this check here with

```
# shellcheck disable=SC2154
profile_file="${ENV_DIR}/${__profile}.env"
```

**Comment by erikw** [comment]
Thanks for the fixes! I "fixed" the linter warnings and the only two small things would be

* `set -euo pipefail`
* possibly error message when can't reading profile

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:4]:
```suggestion
```

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:128]:
```suggestion

# Exit on error, unbound variable, pipe error
set -euo pipefail
```

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:123]:
```suggestion
eval "$(docopt "$@")" # See https://github.com/andsens/docopt.sh for the magic :)
```

File: usr/local/sbin/resticw
Comment (erikw) [inline usr/local/sbin/resticw:140]:
```suggestion
  echo "Error: could not read the environment file ${profile_file}. Are you running as the correct use? Maybe try sudo with the right user.
```

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw]: :facepalm:  oh yes! I didn't think on that solution!! I updated it with the other suggestions

File: usr/local/sbin/resticw
Comment (gerardbosch) [reply usr/local/sbin/resticw]: Sure! I commited this along with other suggestions ;)

**Comment by erikw** [comment]
Alrigh, it's in 🚀 🚀 !

**Comment by gerardbosch** [comment]
Nice!

**Pull request merged**
